Use refpool optimized method for integer grouping

nalimilan opened the pull request:
main:
```julia
julia> using DataFrames, BenchmarkTools

julia> df = DataFrame(x=rand(1:10, 100_000), y=rand(10:20, 100_000));

julia> @btime groupby(df, :x);
  730.118 μs (31 allocations: 2.53 MiB)

julia> @btime groupby(df, [:x, :y]);
  1.091 ms (38 allocations: 2.53 MiB)

julia> df = DataFrame(x=rand(1:1000, 100_000), y=rand(1000:2000, 100_000));

julia> @btime groupby(df, :x);
  1.009 ms (31 allocations: 2.53 MiB)

julia> @btime groupby(df, [:x, :y]);
  2.690 ms (38 allocations: 2.53 MiB)
```

PR (updated to latest commits):
```julia
jjulia> using DataFrames, BenchmarkTools

julia> df = DataFrame(x=rand(1:10, 100_000), y=rand(10:20, 100_000));

julia> @btime groupby(df, :x);
  186.219 μs (58 allocations: 784.46 KiB)

julia> @btime groupby(df, [:x, :y]);
  296.452 μs (85 allocations: 785.84 KiB)

julia> df = DataFrame(x=rand(1:1000, 100_000), y=rand(1000:2000, 100_000));

julia> @btime groupby(df, :x);
  187.802 μs (58 allocations: 785.43 KiB)

# Too many combinations, falling back to generic method
julia> @btime groupby(df, [:x, :y]);
  2.982 ms (98 allocations: 2.53 MiB)
```

nalimilan commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,43 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+    if refpool !== nothing
+        return refpool
+    elseif x isa AbstractArray{<:Real} && !isempty(x) && all(isinteger, x)
+        minval, maxval = extrema(x)
+        # Threshold chosen with the same rationale as the row_group_slots refpool method:
+        # refpool approach is faster but we should not allocate too much memory either
+        if maxval - minval + 1 <= 2 * length(x)
```
It seems that even above this threshold the `refpool` approach is faster. But I'm not sure how much memory we're willing to allocate for speed.

nalimilan commented:
Last commit will fail until JuliaData/Missings.jl#126.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+# this just allows telling the generic row_group_slots method for generic refpools
+# what is the minimum index and the number of (potential) groups
+# missing values are supported and represented used index max+1
+struct IntegerRefpool{T<:Union{Real, Missing}} <: AbstractVector{T}
```
we should disallow bare `Missing` I think

bkamins replied:
or not? (I just started reading this PR 😄 )

bkamins replied:
Also `Union{}` meets the restriction you specified.

nalimilan replied:
I don't think that's a problem. I'll add tests.

bkamins replied:
OK - it is just my second reflex that in such `Union` cases `Missing` and `Union{}` are corner case that should be considered. If it is not a problem here then it is great.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+end
+
+Base.size(x::IntegerRefpool{T}) where {T} = (x.max - x.min + 1 + (T >: Missing),)
+Base.axes(x::IntegerRefpool{T}) where {T} = (x.min:(x.max + (T >: Missing)),)
```
we should check then upon construction that `x.max + 1` does not overflow. The same for `x.max - x.min + 1 + (T >: Missing)`. Alternatively here we can perform `checked_add` here (it is slower than normal one, but it probably does not matter here)

bkamins replied:
same with `cheked_sub`

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+
+    if refpool !== nothing
+        return refpool, refarray
+    elseif x isa AbstractArray{<:Union{Real, Missing}} &&
```
same here we allow `Union{}` and `Missing` eltype. Do we want this?

nalimilan replied:
Yeah, `Missing` is caught by `isempty(skipmissing(x))`, and `Union{}` is either caught by the same check when empty, or triggers an error that will happen anyway when trying to access a value.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+        all(v -> ismissing(v) | isinteger(v), x)
+        isempty(skipmissing(x)) && return nothing, nothing
+        minval, maxval = extrema(skipmissing(x))
+        # Threshold chosen with the same rationale as the row_group_slots refpool method:
```
I think we can leave this threshold for now.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+        # Threshold chosen with the same rationale as the row_group_slots refpool method:
+        # refpool approach is faster but we should not allocate too much memory either
+        if maxval - minval + 1 <= 2 * length(x)
+            refpool′ = IntegerRefpool{eltype(x)}(minval, maxval)
```
we should check here if `minval` and `maxval` can be cast to `Int`. If not we should fall back to a slow path.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+        minval, maxval = extrema(skipmissing(x))
+        # Threshold chosen with the same rationale as the row_group_slots refpool method:
+        # refpool approach is faster but we should not allocate too much memory either
+        if maxval - minval + 1 <= 2 * length(x)
```
`maxval - minval + 1` can overflow.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,59 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+            refpool′ = IntegerRefpool{eltype(x)}(minval, maxval)
+            if eltype(x) >: Missing
+                # Missing values go to last group with code maxval+1
+                refarray′ = Missings.replace(x, maxval+1)
```
again - we should earlier check if `maxval+1` does not overflow `Int`.

bkamins commented:
Looks very nice. We only need to make sure we correctly disable trying this path in corner cases (like `BigInt` vector that has very large values, but with small range) as otherwise we will have overflow.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,68 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+    min::Int
+    max::Int
+    function IntegerRefpool{T}(min::Real, max::Real) where T<:Union{Int, Missing}
+        @assert max < typemax(Int) - 1
+        @assert typemin(Int) <= widen(max) - widen(min) + 1 < typemax(Int)
```
Small suggestions:
1. check that `min` is less of equal than `max`
2. then no need to check `typemin(Int)`
3. also `+2` is needed as we potentially have `Missing` to handle
```suggestion
        @assert min <= max < typemax(Int) - 1
        @assert widen(max) - widen(min) + 2 < typemax(Int)
```

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,68 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+Base.size(x::IntegerRefpool{T}) where {T} = (x.max - x.min + 1 + (T >: Missing),)
+Base.axes(x::IntegerRefpool{T}) where {T} = (x.min:(x.max + (T >: Missing)),)
+Base.IndexStyle(::Type{<:IntegerRefpool}) = Base.IndexLinear()
+@inline function Base.getindex(x::IntegerRefpool{T}, i::Real) where T
```
just to be sure - do we have a test when `i` is e.g. `1.0` or `true` to make sure that we error in this case? In general why do you want to allow `i` to be `Real` rather than `Union{Signed, Unsigned}`? (actually the same question is to line 105 with inner constructor)

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,68 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+
+    if refpool !== nothing
+        return refpool, refarray
+    elseif x isa AbstractArray{<:Union{Real, Missing}} &&
```
again - why do we allow `Real` here?

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,68 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+        # We also have to avoid overflow
+        if typemin(Int) <= maxval + 1 < typemax(Int) &&
+            typemin(Int) <= minval <= typemax(Int) &&
+            widen(maxval) - widen(minval) + 1 <= 2 * length(x) < typemax(Int)
```
as above you have to use `+2` to allow to store `missing`
```suggestion
            widen(maxval) - widen(minval) + 2 <= 2 * length(x) < typemax(Int)
```

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -163,8 +230,12 @@ function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
-                         refpools::NTuple{N, <:AbstractVector},
+function row_group_slots(cols::NTuple{N, AbstractVector},
+                         refpools::NTuple{N, AbstractVector},
+                         refarrays::NTuple{N,
```
These line breaks are ugly, but probably we cannot help it.

bkamins commented:
I would also add a test that on 32-bit machine uses slow path and on 64-bit machine uses fast path (as `Int` is defined differently on both) and another test that on 64-bit machine uses slow path because the values are too large.

nalimilan commented:
I've added checks and tests for overflows. I've also bumped into an already existing bug with `skipmissing=true` when there are only missing values so I've added more tests for that.

bkamins commented:
Have you seen my comments I left today in the morning?

nalimilan commented:
> I would also add a test that on 32-bit machine uses slow path and on 64-bit machine uses fast path (as `Int` is defined differently on both) and another test that on 64-bit machine uses slow path because the values are too large.

I've added tests that the fast path is used just below `typemax(Int)`, and correctness tests just above it.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,84 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+    if eltype(x.x) >: Missing && v === missing
+        return x.replacement
+    else
+        return Int(v - x.offset)
```
are we sure `v - x.offset` will not overflow here?

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,84 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+            ngroups + 1 <= 2 * length(x) <= typemax(Int)
+            T = eltype(x) >: Missing ? Union{Int, Missing} : Int
+            refpool′ = IntegerRefpool{T}(Int(ngroups))
+            refarray′ = IntegerRefarray(x, Int(minval) - 1, Int(ngroups) + 1)
```
OK - I guess this line answers my question above. If we initialize `offset` to what you propose then we are safe. Maybe just add a comment above what `offset` means for `IntegerRefarray` type?

nalimilan replied:
I've added a comment.

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,84 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+        # (note that it would be possible to allow minval and maxval to be outside of the
+        # range supported by Int by adding a type parameter for minval to IntegerRefarray)
+        if typemin(Int) < minval <= maxval < typemax(Int) &&
+            ngroups + 1 <= 2 * length(x) <= typemax(Int)
```
it is interesting here that `2*length(x)` can overflow if we are on 32 bit machine, but then it will be negative so the condition is correct (simply we will not use a fast path then). But if you want to be correct maybe better write `Int64(2) * length(x)`?

nalimilan replied:
Woops, that's indeed not super clean. I've switched to `Int64` (including in two similar occurrences).

bkamins commented on src/dataframerow/utils.jl:
```diff
@@ -94,6 +94,84 @@ isequal_row(cols1::Tuple{Vararg{AbstractVector}}, r1::Int,
+    if refpool !== nothing
+        return refpool, refarray
+    elseif x isa AbstractArray{<:Union{Real, Missing}} &&
+        all(v -> ismissing(v) | isinteger(v), x) &&
```
Have you checked that using `|` here is is optimized by the compiler correctly? What I mean is that when `eltype(x) <: Integer` is this check a no-op?

nalimilan replied:
Yes, that's a good point, but I had tested it and it works (which is really nice).

bkamins commented:
Looks good. Thank you. Can you just please make sure that in the points I have just raised we are clear what happens?

bkamins commented:
Thank you. Looks good now. These overflow things are tricky.
